### PR TITLE
JIT: Factor SSA's DFS and profile synthesis's loop finding

### PR DESCRIPTION
**Pull request by jakobbotsch**
Factor out SSA's general DFS (that takes EH into account) and encapsulate it in a `FlowGraphDfsTree` class.

Factor out profile synthesis's loop finding and encapsulate it in a `FlowGraphNaturalLoops` class. Switch construction of it to use the general DFS instead of the restricted one (that does not account for exceptional flow).

Optimize a few things in the process:
* Avoid storing loop blocks in a larger than necessary bit vector; store them starting from the loop header's postorder index instead.
* Provide post-order and reverse post-order visitors for the loop blocks; switch profile synthesis to use this in a place

No diffs are expected. A small amount of diffs are expected when profile synthesis is enabled due to the modelling of exceptional flow and also from handling unreachable predecessors (which would reject some loops as unnatural loops before).

My future plans are to proceed to replace the loop representation of loops with this factored version, removing the lexicality requirement in the process, and hopefully fixing some of our deficiencies.

### CODE REVIEW

File: src/coreclr/jit/flowgraph.cpp [hunk 1]
```diff
@@ -4083,3 +4083,560 @@ void Compiler::fgLclFldAssign(unsigned lclNum)
...
+
+    // Work back through flow to loop head or to another pred
+    // that is clearly outside the loop.
+    //
+    while (!worklist.empty())
+    {
+        BasicBlock* const loopBlock = worklist.back();
+        worklist.pop_back();
+
+        for (FlowEdge* predEdge = comp->BlockPredsWithEH(loopBlock); predEdge != nullptr;
```
Comment (jakobbotsch) [inline src/coreclr/jit/flowgraph.cpp:4616]: This one is the important one to switch to `BlockPredsWithEH` to recognize the full SCC of loops containing EH constructs.

File: src/coreclr/jit/flowgraph.cpp [hunk 2]
```diff
@@ -4083,3 +4083,560 @@ void Compiler::fgLclFldAssign(unsigned lclNum)
...
+            loops->m_improperLoopHeaders++;
+            continue;
+        }
+
+        JITDUMP("Loop has %u blocks\n", BitVecOps::Count(&loopTraits, loop->m_blocks));
+
+        // Find the exit edges
+        //
+        loop->VisitLoopBlocks([=](BasicBlock* loopBlock) {
+            loopBlock->VisitRegularSuccs(comp, [=](BasicBlock* succBlock) {
```
[the remark this answers is not in the document]
Comment (AndyAyersMS) [reply src/coreclr/jit/flowgraph.cpp:4480]: Yeah, makes sense -- for profile purposes we will assume exceptions never happen.

**Review by AndyAyersMS — APPROVED** [review]
LGTM.

**Comment by jakobbotsch** [comment]
I'm going to merge this, but @BruceForstall please feel free to take a look when you have time and I'll try to address your feedback in a follow-up.

**Pull request merged**